### ISSUE TITLE
Open libraries count as real files when opened via CVMFS

### ISSUE THREAD

**Issue by ktf** [cvmfs/cvmfs#3067]
When a library is loaded from disk, it does not count against the open file descriptors because it gets opened, mmaped, and then the file descriptor is closed.

This is not the case when loading a library in CVMFS. What happens is that the CVMFS fuse daemon keeps an open fd to the file in the cache, even after the file has been closed on the user side. This happens on a per process basis, so that with an handful of processes loading a few library you could have a severe impact on the amount of open file descriptors for the cvmfs deamon, which itself is limited by the CVMFS_NFILE limit.

Is there any particular reason why that needs to be the case? Why can't the fuse daemon simply close the file when the close syscall is issued by dlopen?

**Comment by jblomer** [comment]
That's mandated by the semantics of the Linux file system layer. For the file system, it doesn't matter if the process has the file open for reading or mmaped the file. In both cases, the inode is active.

We currently set a 128k limit for file descriptors in cvmfs. Did you hit that limit?

**Comment by ktf** [comment]
Ciao, yes, we hit that limit because we have a number of processes and the .pcm files of root or shared libraries all count as one, while on other filesystems they get multiplexed. I can understand that 1 inode is active per file, but why one per process? After all they are all opened as readonly.

**Comment by jblomer** [comment]
I see, thanks for the additional information!

The fastest fix, when you control the client, is to just increase `CVMFS_NFILES`. It should be safe to go as high as 1M. We can also generally discuss increasing the default limit (not sure how much you typically overshoot our current default).

Regarding multiplexing: that happens for the memory mappings, i.e. every mmaped file (inode) consumes memory only once even if read by multiple processes.  For file descriptors, the cvmfs fuse module (process) reads all data from the local disk cache, and to do so it "duplicates" all file descriptors of applications using cvmfs.  Native file systems just access block data directly.

It may be possible to re-use the file descriptor for a file already open in the local cache. That would only help for non-chunked files though, and I expect many of the C++ libraries are "large" (> few MB) and therefore chunked.  For chunked files, different process may ready different parts of the file at a given moment, so they each trigger another chunk to be opened.

One could also consider that cvmfs closes file descriptors when it does not currently read from them. This is tricky to get right though, e.g. when should a descriptor be closed, what happens if re-opening fails etc.

Both ideas to reduce the number of open file descriptors in cvmfs are not quick fixes. Can you give me a bit more context about when you see the problem, what will be your limit on the processes & open files and how much sharing of files you expect to see?

**Comment by ktf** [comment]
Doubling the current amount seems to be enough for us.

When you say "chunks" you mean segments? So basically each library is at least 4 inodes, correct?

What happens on fork + exec? Do you close the inodes for the parent?

**Comment by jblomer** [comment]
Chunks are calculated based on file size and a rolling checksum.  The alice.cern.ch uses the default parameters, i.e. 8MB average chunk size, 4MB minimum and 16MB maximum. The algorithm does not look into the contents of the file, so the chunks are not aligned with segment boundaries.

For fork and exec, it depends if `FD_CLOEXEC` is used by the parent. Cvmfs will strictly track the open file descriptors: if both parent and child have an fd on /cvmfs open or mmap a file from /cvmfs, they will trigger 2 open files in the fuse module.

**Comment by traylenator** [comment]
For info at CERN open files now doubled:

Higest alice number we are seeing is 169181 which inside the new max of 262144.

**Comment by ktf** [comment]
@traylenator thank you. Could you provide me a link to the monitoring, so that I can check how much the changes I am doing impact production?

**Comment by jblomer** [comment]
We have been seeing now the ALICE software framework in action, and confirmed that all processes map more or less the same shared libraries but trigger each an extra copy of open file descriptors in the fuse module.  Next steps:

1. Increase `CVMFS_NFILES` for alice.cern.ch in the config repository
2. Collect statistics from lxplus/lxbatch on the number of duplicate file opens in the cvmfs cache. @traylenator Would it be possible to get a selection of `ls -l /proc/$(attr -qg pid /cvmfs/<reponame>)/fd` for alice.cern.ch, cms.cern.ch, and atlas.cern.ch from lxplus?
3. Investigate coalescing of file descriptors in the cvmfs posix cache manager. We'd need to introduce a new, thread-safe table `object hash --> fd + refcount` that needs to survive reloads/hotpatches. On the plus side, we would save `open()` and `close()` calls.

**Comment by jblomer** [comment]
See https://github.com/cvmfs-contrib/config-repo/pull/121 for the change to `CVMFS_NFILES`

**Comment by traylenator** [comment]
If I understand our ACLs correctly which I don't @ktf , @jblomer you should both have access to:

https://monit-grafana.cern.ch/d/QjvP5Y74z/cvmfs-dashboard?orgId=5

contact me offline if you don't.

**Comment by olifre** [comment]
I found an "easy" reproducer exceeding the current 128k file limit: Trying to compile ROOT on a machine with 256 logical cores, using `make -j256`, and using a singularity / apptainer container image from CVMFS. 

In the initial phase which copies over the source files, runs small python helpers and g++ with many includes, this easily exceeds 128k (in a measurement every 0.1 seconds, I saw `188340` open FDs), tested in my case with a Debian 11 container image and CVMFS 2.10.1.

**Comment by vvolkl** [comment]
Hi @olifre,

thanks for testing this - we've just built the packages for CVMFS 2.11.0, (which will be announced shortly on cvmfs-testing@cern.ch). Could you perhaps repeat your test with 2.11.0 ( from the testing repository, or take directly the `cvmfs` and `cvmfs-libs` packages from http://ecsft.cern.ch/dist/cvmfs/cvmfs-2.11.0/ ) and the `CVMFS_CACHE_REFCOUNT=yes` option enabled?
Cheers,
Valentin

**Comment by olifre** [comment]
Hi @vvolkl ,

many thanks, great timing!
I updated directly from the RPMs, explicitly unmounted and remounted to force all `cmvfs2` processes to restart completely. 

I use:
```
while true; do ls -l /proc/$(attr -qg pid /cvmfs/container-repo)/fd 2>/dev/null| wc -l; sleep 0.1; done
```
for testing — of course, this poor-mans-monitoring yields a big error band, trying to get the peak of open FDs this way is quite "uinrelaible" on such a fast node, but it should be fine to get the order of magnitude. 

In two tests (without `CVMFS_CACHE_REFCOUNT`) I got `138788` and `169772`, so without the option, CVMFS 2.11.0 behaves similarly (as should be expected). 

With `CVMFS_CACHE_REFCOUNT`, the difference is (positively) dramatic! 

In the first test, I get about `345` open FDs at roughly the same place during ROOTs compilation (i.e. the copying / python script preparation steps), and about `580` later on when actual compilation kicks in. 
In a second run, I got about `780` open FDs for the early phase, which is likely just since the Python scripts and copying processes are racing slightly non-deterministically against each other. 

So yes, that is a very dramatic positive change, in a very good way :+1: . 

Cheers,
Oliver

**Comment by maarten-litmaath** [comment]
@traylenator I see 2.11.0 already on QA nodes, but without `CVMFS_CACHE_REFCOUNT=yes` --> can we also have that setting enabled on the QA WNs?  Is it already?

@jblomer Will that setting become the default in the next (which?) release if all looks OK in production?

**Comment by vvolkl** [comment]
This setting will become the default in the next release, but we're planning to first enable this for selected repositories via our cvmfs-config.cern.ch repository in the next few weeks.

**Comment by ktf** [comment]
For the record, I tested with some simple ALICE workflow, and I see a 3x improvement in the number of open file descriptors.

**Comment by maarten-litmaath** [comment]
@vvolkl If the extra setting were enabled for alice.cern.ch, we would get it for ALICE jobs at all sites running >= 2.11.0, which is the goal, but I can imagine we would rather want to test this at **scale** first in QA?

**Comment by traylenator** [comment]
> @traylenator I see 2.11.0 already on QA nodes, but without `CVMFS_CACHE_REFCOUNT=yes` --> can we also have that setting enabled on the QA WNs? Is it already?
> 
> @jblomer Will that setting become the default in the next (which?) release if all looks OK in production?

https://github.com/voxpupuli/puppet-cvmfs/pull/186  has the puppet changes after which time can QA it at CERN.

**Comment by maarten-litmaath** [comment]
@traylenator That looks great, thanks!  Please let us know when it is active in QA.  Is there some monitoring of CVMFS behavior on the WNs that would allow us to see the effects in QA vs. prod?

**Comment by rptaylor** [comment]
I suppose there should be documentation on CVMFS_CACHE_REFCOUNT in https://cvmfs.readthedocs.io/en/latest/apx-parameters.html  ?
Actually CVMFS_STREAMING_CACHE is also mentioned in the 2.11 release notes but not documented AFAICT.

**Comment by vvolkl** [comment]
> I suppose there should be documentation on CVMFS_CACHE_REFCOUNT in https://cvmfs.readthedocs.io/en/latest/apx-parameters.html ? Actually CVMFS_STREAMING_CACHE is also mentioned in the 2.11 release notes but not documented AFAICT.

I think the docs were there in the 2.11 and stable version, just missing in the "latest" version. Updated now.

**Comment by traylenator** [comment]
`CVMFS_CACHE_REFCOUNT=yes` is now deploying to batchtest (1%).

The link above to the plot has an `alice` and `batchtest` drop down which should change for open files in principal....

Will look tomorrow.

**Comment by ktf** [comment]
For the record, I now get an "access denied" when trying to open the link above (https://monit-grafana.cern.ch/d/QjvP5Y74z/cvmfs-dashboard?orgId=5)

**Comment by maarten-litmaath** [comment]
That link may just be accessible for the IT Department. Anyway, I suppose a before-and-after plot should be uploaded here for the record...

**Comment by traylenator** [comment]
![image](https://github.com/cvmfs/cvmfs/assets/135205/9492c9ec-344c-4726-aa68-0e60c7916cc8)

Average and max open files for `alice.cern.ch` in batchops environment for the last 7 days. 

Change was made on Morning of Wednesday 6th of September - maybe an improvement but not very convincing.

**Comment by ktf** [comment]
Is this only for the nodes in QA?

**Comment by olifre** [comment]
@traylenator Did you "only" set the option via Puppet, or perform a full remount? 

A difference I see between my earlier test in this issue and yours could be that I fully remounted all repositories on my test node. Since the refcounting happens at quite a low level in CVMFS (i believe), maybe `cvmfs_config reload` is insufficient to put the change live?

**Comment by olifre** [comment]
At least the test at:
https://github.com/vvolkl/cvmfs/blob/e5082d7af263e87f9d7393437b8b7befbaf1efda/test/src/103-reloadcachemgr/main
indicates that `cvmfs_config reload` should be sufficient, and it should be possible to confirm this with:
```
cvmfs_talk cache instance
````
So it may be interesting to check whether this responds as expected on the `batchtest` nodes?

**Comment by jblomer** [comment]
The recounted cache can be activated by a reload. It will only reduce file descriptor use among the files opened after the reload though.  So the effect without a full remount cycle will probably only show slowly over time.

**Comment by traylenator** [comment]
> batchops

Yes - `batchtest` yes . all nodes in these plots should have new config and a reload issued.

**Comment by ktf** [comment]
Can we have an updated plot, then, given the point of @jblomer above?

**Comment by traylenator** [comment]
> Can we have an updated plot, then, given the point of @jblomer above?

Not really as only 7 days of retention.

Looking at a high recent node:

```
2023-09-11 14:37:40 bw.example.ch alice.cern.ch 185113
2023-09-11 13:37:40 bw.example.ch alice.cern.ch 185113
2023-09-11 12:37:40 bw.example.ch alice.cern.ch 185113
2023-09-11 04:37:40 bw.example.ch alice.cern.ch 159310
2023-09-11 03:37:40 bw.example.ch alice.cern.ch 159310 
2023-09-11 05:37:40 bw.example.ch alice.cern.ch 158991
```
This is configured correctly:

```
[root@bw ~]# grep REF /etc/cvmfs/default.local 
CVMFS_CACHE_REFCOUNT='yes'
```
The mounts are recent:

```
[root@bw ~]# ps auxwww | egrep 'cachemgr|cvmfs2'
cvmfs     978987  0.0  0.0  16400   144 ?        S    12:21   0:00 /usr/bin/cvmfs2 -o rw,system_mount,fsname=cvmfs2,allow_other,grab_mountpoint,uid=990,gid=989 cvmfs-config.cern.ch /cvmfs/cvmfs-config.cern.ch
cvmfs     978991  0.4  0.0  29268  6648 ?        Sl   12:21   0:45 /usr/bin/cvmfs2 __cachemgr__ . 9 10 362396254208 181198127104 0 3 -1 :
cvmfs     978995  0.0  0.0  16280   140 ?        S    12:21   0:00 /usr/bin/cvmfs2 __cachemgr__ . 9 10 362396254208 181198127104 0 3 -1 :
cvmfs     979000  0.0  0.0 609824 26968 ?        Sl   12:21   0:00 /usr/bin/cvmfs2 -o rw,system_mount,fsname=cvmfs2,allow_other,grab_mountpoint,uid=990,gid=989 cvmfs-config.cern.ch /cvmfs/cvmfs-config.cern.ch
cvmfs     979075  0.0  0.0  16404   180 ?        S    12:21   0:00 /usr/bin/cvmfs2 -o rw,system_mount,fsname=cvmfs2,allow_other,grab_mountpoint,uid=990,gid=989 alice.cern.ch /cvmfs/alice.cern.ch
cvmfs     979079  6.8  0.0 17768260 190156 ?     Sl   12:21  12:31 /usr/bin/cvmfs2 -o rw,system_mount,fsname=cvmfs2,allow_other,grab_mountpoint,uid=990,gid=989 alice.cern.ch /cvmfs/alice.cern.ch
```

**Comment by ktf** [comment]
Could we have a list of those open files put somewhere?

**Comment by DrDaveD** [comment]
Does this fix only apply to cvmfs clients, not to snapshots?  (KEK at the moment is having a big problem with using too many file descriptors when snapshotting unpacked.cern.ch).

**Comment by vvolkl** [comment]
> Does this fix only apply to cvmfs clients, not to snapshots? (KEK at the moment is having a big problem with using too many file descriptors when snapshotting unpacked.cern.ch).

No, as far as I can tell snapshotting does not use the client cache unfortunately. We'll look into this separately.

**Comment by vvolkl** [comment]
@traylenator Could you share another fd monitoring update?

**Comment by jblomer** [comment]
@traylenator It would be useful if we can see `/proc/<cvmfs2 fuse process id>/fd`
Just to verify that all these open files are distinct and the refcounting mechanism works as expected.

**Comment by traylenator** [comment]
[output.txt](https://github.com/cvmfs/cvmfs/files/12869246/output.txt)

Sorry for delay - this is from a :
```
/usr/bin/cvmfs2 -o rw,system_mount,fsname=cvmfs2,allow_other,grab_mountpoint,uid=990,gid=989 alice.cern.ch /cvmfs/alice.cern.ch
```

**Comment by jblomer** [comment]
Thanks a lot! This output suggests that the fd deduplication does not work as expected. While we see it working in our tests, here it clearly shows that the same files are open multiple times. It seems to be more complicated than a simple race, e.g. looking at the file /pool/cvmfs/shared/f7/6d546b68c8a7fc3e888144c5928040601b5700 the corresponding fds were opened during a window of ~1h.

Just to be sure, @traylenator could you run for this very instance `cvmfs_talk -i alice.cern.ch cache instance` to verify that the recounted cache is at work? And perhaps also `cvmfs_talk -i alice.cern.ch cleanup rate 1500` to see if the cache is thrashing.

**Comment by traylenator** [comment]
You were correct to question those results - forget was only applied to particular subset - please disregard that one.


New results on a node currently sitting high for alice.

```
# cvmfs_talk -i alice.cern.ch cache instance
Refcounting Posix cache manager(cache directory: .)
# cvmfs_talk -i alice.cern.ch cleanup rate 1500
0
```
and ...

[output.txt](https://github.com/cvmfs/cvmfs/files/12869745/output.txt)

**Comment by vvolkl** [comment]
Thanks! In the latest output every cache file has only one  file descriptor, which is what we wanted the new cache manager to do. However we'd also expect this to have a measureable impact on the system. If this is not the case on the batch, I think we'll have to do  more digging to find out why - in every test we did there was a significant reduction in the number of fds for parallel reads.

**Comment by traylenator** [comment]
![image](https://github.com/cvmfs/cvmfs/assets/135205/8fcf9240-9bde-4481-8e71-ebe7ff251ffe)

Here's a fresh plot for the last 7 days and alice - `batchtest` nodes only = refcount ones.

**Comment by ktf** [comment]
Also, in the latest report you have ~1000 files from CVMFS, what are the others?

**Comment by ktf** [comment]
Ping? Can we have a report from when the file count is high? What is the grafana looking at, BTW? How is it populated? Are we sure the plot themselves don't do the double counting?

**Comment by traylenator** [comment]
Getting the instantaneous values now:

```
$ mco rpc shell run command="ls /cvmfs/alice.cern.ch > /dev/null && attr -q -g usedfd /cvmfs/alice.cern.ch" -T bi --dt 10 -F environment=batchtest -T bi -F environment=batchtest --json | tee results.json
```

```
$ cat results.json | jq -r  .[].data.stdout | sort -nr | head -20
108952
104745
102152
97141
84334
80844
67725
66688
62778
```

Plotting just the last 30 minutes:

![image](https://github.com/cvmfs/cvmfs/assets/135205/dcad85ee-6a7a-479d-a596-db7746861945)

so only one data point they look to match.

May it's worth adding a `max_usedfd` ?

**Comment by traylenator** [comment]
Will try and catch a list of open file for when we are at the limit.... Thinking....

**Comment by jblomer** [comment]
Ah, that's something we forgot!  

Our internal accounting does not take into account the refcounting.  To get the actual number of open file descriptors, you'd currently need to look into `/proc/<pid of cvmfs2 fuse process>/fd`.

Unfortunately, that also means we have to keep `CVMFS_NFILES` parameter high because otherwise the cvmfs code returns `EMFILE` even though in reality there are still plenty of file descriptors available.

We'll fix our accounting for the 2.11.2 patch release.

**Comment by olifre** [comment]
Hurray, that finally clears up why I saw the huge improvement in my tests (I used the "poor man's monitoring" as outlined in https://github.com/cvmfs/cvmfs/issues/3067#issuecomment-1681240722 and not the internal accounting :wink: ).

**Comment by ktf** [comment]
Lies, big lies, and monitoring systems. ;-)

**Comment by traylenator** [comment]
Have added code to enable something to snapshot FD when we trip `usedfd` but from comments above I will not bother to enable it.

**Comment by DrDaveD** [comment]
Is this now solved by #3426?

**Comment by traylenator** [comment]
Running:
```
ls /cvmfs/alice.cern.ch > /dev/null ; ls -1 /proc/$(cvmfs_talk -i alice.cern.ch pid)/fd | wc -l
```
across the batch tests nodes now the highest values are:

```
3965
4243
4325
4999
8078
15358
32158
```
Taking a look at that large one "32158"

```
# readlink /proc/$(cvmfs_talk -i alice.cern.ch pid)/fd/* | sort | uniq  -c | sort -n | tail
      2 pipe:[1092798206]
      2 pipe:[1092798207]
      2 pipe:[1092798208]
      2 pipe:[1092798209]
      2 pipe:[1092798210]
      2 pipe:[1092798211]
      2 pipe:[1092798212]
      2 /pool/cvmfs/shared/pipe0 (deleted)
      3 /dev/null
  30495 /dev/fuse
  ```
which is probably interesting but unrelated - will look at that offline with cvmfs folk.

As expected the `attr -g usedfd` values much higher than they should be still.

**Comment by olifre** [comment]
> As expected the `attr -g usedfd` values much higher than they should be still.

I have not done tests on my end yet (currently on vacation), but does 2.11.2 fix also the internal accounting for you, i.e. does the monitoring now also show the improvements? 
If so, all issues mentioned here are fixed.

**Comment by vvolkl** [comment]
> Is this now solved by #3426?

@DrDaveD I'd say yes - the internal accounting with `attr -g usedfd` cannot yet be used with the reference-counting cache manager mode, but  that's a separate issue, and needs some restructuring in the code, so will only go in the next feature release. @ktf Can we close this?

**Comment by maarten-litmaath** [comment]
Hi all,
what is the right value for `CVMFS_CACHE_REFCOUNT` to enable it?
The documentation says `yes`, but I have also seen `true` and even `1`:
are they all equivalent?  Is that documented somewhere?

**Comment by vvolkl** [comment]
All boolean CVMFS parameters can be set to either `yes`/`no`, `true`/`false`, `on`/`off` or `1`/`0`, These are all equivalent and case-insensitive. I think the docs don't mention this and mostly talk about setting things to `yes` or `no`, so that's a good point, I'll add a note.

**Comment by ktf** [comment]
@traylenator Can we have a new plot with and without the change, showing the improvement on the testbed?

Apart from from that, I have no objections closing this.

**Comment by traylenator** [comment]
> @traylenator Can we have a new plot with and without the change, showing the improvement on the testbed?

No since as above 

> the internal accounting with attr -g usedfd cannot yet be used with the reference-counting cache manager mode

i.e these numbers have not changed yet  which is what I was and am plotting.

**Comment by ktf** [comment]
I meant whenever you have the new metric in place. Also, no need to wait for that metric to be there to close this, if you state you are officially happy with the change and that it solves the concerns about too many open file descriptors by ALICE.

**Comment by traylenator** [comment]
For sure on inspection it's clearly massively lower than was before.

**Comment by ktf** [comment]
Good enough, feel free to close as fixed.

Ciao,
Giulio
